src/ripple/conditions/impl/Der.h

```diff
@@ -0,0 +1,2798 @@
+            auto const c = static_cast<signed char>((v >> (n * 8)) & 0xff);
+            if (c == toSkip &&
+                !(isSigned && v < 0 && n &&
+                  (static_cast<signed char>((v >> ((n - 1) * 8)) & 0xff) >= 0)))
+                continue;
+
+            if (v > 0 && c < 0)
+                return n + 2;
+            else
+                return n + 1;
+        }
+        assert(0);  // will never happen
+        return 1;
+    }
+
+    template <class T>
+    static
+    std::uint64_t
+    length(
+        T const& v,
+        boost::optional<GroupType> const& parentGroupType,
+        TagMode encoderTagMode,
+        TraitsCache& traitsCache)
+    {
+        return length(v);
+    }
+
+    template <class T>
+    static
+    void
+    encode(Encoder& s, T v)
+    {
+        if (s.subgroups_.empty())
+        {
+            s.ec_ = make_error_code(Error::logicError);
+            return;
+        }
+
+        auto& parentSlice = s.parentSlice();
+
+        if (!v)
+        {
+            if (parentSlice.empty())
+            {
+                s.ec_ = make_error_code(Error::logicError);
+                return;
+            }
+            parentSlice.push_back(0);
+            return;
+        }
+
+        boost::optional<GroupType> parentGroupType;
+        if (!s.subgroups_.empty())
+            parentGroupType.emplace(s.subgroups_.top().groupType());
+        std::size_t n = length(v, parentGroupType, s.tagMode_, s.traitsCache_);
+        if (parentSlice.size() != n)
+        {
+            s.ec_ = make_error_code(Error::logicError);
+            return;
+        }
+        while (n--)
+        {
+            if (n >= sizeof(T))
+                parentSlice.push_back(static_cast<char>(0));
+            else
+                parentSlice.push_back(static_cast<char>((v >> (n * 8)) & 0xFF));
+        }
+    }
+
+    template <class T>
+    static
+    void
+    decode(Decoder& decoder, T& v)
+    {
+        auto& slice = decoder.parentSlice();
+        std::error_code& ec = decoder.ec_;
+
+        if (slice.empty())
+        {
+            // can never have zero sized integers
+            ec = make_error_code(Error::contentLengthMismatch);
+            return;
+        }
+
+        const bool isSigned = std::numeric_limits<T>::is_signed;
+        // unsigned types may have a leading zero octet
+        const size_t maxLength = isSigned ? sizeof(T) : sizeof(T) + 1;
+        if (slice.size() > maxLength)
+        {
+            ec = make_error_code(Error::integerBounds);
+            return;
+        }
+
+        if (!isSigned && (slice[0] & (1 << 7)))
+        {
+            // trying to decode a negative number into a positive value
+            ec = make_error_code(Error::integerBounds);
+            return;
+        }
+
+        if (!isSigned && slice.size() == sizeof(T) + 1 && slice[0])
+        {
+            // since integers are coded as two's complement, the first byte may
+            // be zero for unsigned reps
```

Comment:
No, it may be zero, but it is not necessarily zero. For example, the unsigned value one is encoded as 0x1. If the leading 8th bit of a positive value is 1, we need to add a leading 0 to distinguish it from a negative value.